pwnlib/windbg.py

```diff
@@ -0,0 +1,431 @@
+# -*- coding: utf-8 -*-
+"""
+During exploit development, it is frequently useful to debug the
+target binary under WinDbg.
+
+Pwntools makes this easy-to-do with a handful of helper routines, designed
+to make your exploit-debug-update cycles much faster.
+
+Useful Functions
+----------------
+
+- :func:`attach` - Attach to an existing process
+- :func:`debug` - Start a new process under a debugger, stopped at the first instruction
+- :func:`debug_shellcode` - Build a binary with the provided shellcode, and start it under a debugger
+
+Debugging Tips
+--------------
+
+The :func:`attach` and :func:`debug` functions will likely be your bread and
+butter for debugging.
+
+Both allow you to provide a script to pass to GDB when it is started, so that
+it can automatically set your breakpoints.
+
+Attaching to Processes
+~~~~~~~~~~~~~~~~~~~~~~
+
+To attach to an existing process, just use :func:`attach`.  It is surprisingly
+versatile, and can attach to a :class:`.process` for simple
+binaries, or will automatically find the correct process to attach to for a
+forking server, if given a :class:`.remote` object.
+
+Spawning New Processes
+~~~~~~~~~~~~~~~~~~~~~~
+
+Attaching to processes with :func:`attach` is useful, but the state the process
+is in may vary.  If you need to attach to a process very early, and debug it from
+the very first instruction (or even the start of ``main``), you instead should use
+:func:`debug`.
+
+When you use :func:`debug`, the return value is a :class:`.tube` object
+that you interact with exactly like normal.
+
+Tips and Troubleshooting
+------------------------
+
+``NOPTRACE`` magic argument
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It's quite cumbersom to comment and un-comment lines containing `attach`.
+
+You can cause these lines to be a no-op by running your script with the
+``NOPTRACE`` argument appended, or with ``PWNLIB_NOPTRACE=1`` in the environment.
+
+::
+
+    $ python exploit.py NOPTRACE
+    [+] Starting local process '/bin/bash': Done
+    [!] Skipping debug attach since context.noptrace==True
+    ...
+
+Member Documentation
+===============================
+"""
+from __future__ import absolute_import
+
+import os
+import random
+import re
+import shlex
+import tempfile
+import time
+import subprocess
+
+from pwnlib import atexit
+from pwnlib import tubes
+from pwnlib.context import LocalContext
+from pwnlib.context import context
+from pwnlib.log import getLogger
+from pwnlib.util import misc
+from pwnlib.util import proc
+
+log = getLogger(__name__)
+
+# @LocalContext
+def debug(args, windbgscript=None, exe=None, env=None ):
+    # """debug(args) -> tube
+
+    # Launch a process in a suspended state, attach debugger and return process
+
+    # Arguments:
```

Comment:
These comments don't conform to Sphinx-compatible / Google Python Docstring format.